test(docs-infra): add missing tests for observables and promises

sonukapoor opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [X] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #34398

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

gkalpak commented:
Marking this as blocked on #34327, since it builds on top of that.

sonukapoor commented:
> Marking this as blocked on #34327, since it builds on top of that.

@gkalpak This is unblocked now since #34327 has been merged. Can you review it please?

gkalpak reviewed (changes requested):
Thx, @sonukapoor ❤️
I've left some comments. Also, the current commit message ("tests for observables and promises") is too generic. Can you, please, change it to be more specific. E.g. something like:

> test(docs-infra): add tests for \`comparing-observables\` example

gkalpak commented:
I, also, just realized that we need to make sure the tests are run on CI (currently they are not afaict) :grin:

mary-poppins commented:
You can preview 583ad7f at https://pr34537-583ad7f.ngbuilds.io/.
You can preview aa9a17b at https://pr34537-aa9a17b.ngbuilds.io/.

gkalpak reviewed (changes requested):
Just few more (minor) comments to go :smiley:

Also, as mentioned in https://github.com/angular/angular/pull/34537#issuecomment-577790886, we need to run these tests on CI. This is similar to what you did for other Observables-related examples in #34063. I.e. you need to:

- Add a dummy `e2e/src/app.e2e-spec.ts` (e.g. similar to [observables/e2e/src/app.e2e-spec.ts][1]).
- Add an appropriate `example-config.json` (e.g. similar to [observables/example-config.json][2] but specifying a different command that actually runs the unit tests).

[1]: https://github.com/sonukapoor/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/aio/content/examples/observables/e2e/src/app.e2e-spec.ts
[2]: https://github.com/sonukapoor/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/aio/content/examples/observables/example-config.json

sonukapoor commented:
> Just few more (minor) comments to go 😃
> 
> Also, as mentioned in [#34537 (comment)](https://github.com/angular/angular/pull/34537#issuecomment-577790886), we need to run these tests on CI. This is similar to what you did for other Observables-related examples in #34063. I.e. you need to:
> 
> * Add a dummy `e2e/src/app.e2e-spec.ts` (e.g. similar to [observables/e2e/src/app.e2e-spec.ts](https://github.com/sonukapoor/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/aio/content/examples/observables/e2e/src/app.e2e-spec.ts)).
> * Add an appropriate `example-config.json` (e.g. similar to [observables/example-config.json](https://github.com/sonukapoor/angular/blob/e0ad9ecda0b8a541b405d2ab35335b90ceb21fd1/aio/content/examples/observables/example-config.json) but specifying a different command that actually runs the unit tests).

Yes, I haven't gotten to that yet. I wanted to make sure that we are happy with the codebase now. I should be able to look into the testing from CI next.

mary-poppins commented:
You can preview d902c0d at https://pr34537-d902c0d.ngbuilds.io/.

mary-poppins commented:
You can preview b5767ea at https://pr34537-b5767ea.ngbuilds.io/.

mary-poppins commented:
You can preview ddfb0fc at https://pr34537-ddfb0fc.ngbuilds.io/.

mary-poppins commented:
You can preview a5231d7 at https://pr34537-a5231d7.ngbuilds.io/.

mary-poppins commented:
You can preview d2bd1cc at https://pr34537-d2bd1cc.ngbuilds.io/.

mary-poppins commented:
You can preview 98a1e73 at https://pr34537-98a1e73.ngbuilds.io/.

mary-poppins commented:
You can preview f8e9e85 at https://pr34537-f8e9e85.ngbuilds.io/.

mary-poppins commented:
You can preview 0c90f43 at https://pr34537-0c90f43.ngbuilds.io/.

mary-poppins commented:
You can preview a5c3bb7 at https://pr34537-a5c3bb7.ngbuilds.io/.

mary-poppins commented:
You can preview 19d6b3b at https://pr34537-19d6b3b.ngbuilds.io/.

sonukapoor commented:
@gkalpak can you review the changes? I have my doubts with the tests. Please see my offline comments.

mary-poppins commented:
You can preview 8c74ce7 at https://pr34537-8c74ce7.ngbuilds.io/.

mary-poppins commented:
You can preview ff3b05c at https://pr34537-ff3b05c.ngbuilds.io/.

mary-poppins commented:
You can preview 66e82fd at https://pr34537-66e82fd.ngbuilds.io/.

mary-poppins commented:
You can preview 958cfbb at https://pr34537-958cfbb.ngbuilds.io/.

mary-poppins commented:
You can preview f481b50 at https://pr34537-f481b50.ngbuilds.io/.

mary-poppins commented:
You can preview 8dd3508 at https://pr34537-8dd3508.ngbuilds.io/.

mary-poppins commented:
You can preview a42e11f at https://pr34537-a42e11f.ngbuilds.io/.

mary-poppins commented:
You can preview f686921 at https://pr34537-f686921.ngbuilds.io/.

mary-poppins commented:
You can preview 8be1f4a at https://pr34537-8be1f4a.ngbuilds.io/.

mary-poppins commented:
You can preview 6bd0beb at https://pr34537-6bd0beb.ngbuilds.io/.

mary-poppins commented:
You can preview 1a9cb83 at https://pr34537-1a9cb83.ngbuilds.io/.

mary-poppins commented:
You can preview 188f002 at https://pr34537-188f002.ngbuilds.io/.

mary-poppins commented:
You can preview da228c3 at https://pr34537-da228c3.ngbuilds.io/.

mary-poppins commented:
You can preview ca6f557 at https://pr34537-ca6f557.ngbuilds.io/.

mary-poppins commented:
You can preview f54befb at https://pr34537-f54befb.ngbuilds.io/.

mary-poppins commented:
You can preview d9391c2 at https://pr34537-d9391c2.ngbuilds.io/.

mary-poppins commented:
You can preview f22a749 at https://pr34537-f22a749.ngbuilds.io/.

sonukapoor commented:
@gkalpak I have fixed all issues you have mentioned, except I am not able to run the tests in the CI environment. I have tried everything I could. Here are the changes I did:

- Added a `tsconfig-custom.app.json`, to transpile the spec files into the `out-tsc` folder. 
- I was able to see that the files are generated fine by temporarily adding a `js` file that prints out the contents of that folder. 
- I also added a `jasmine.json` file as config for `jasmine`. 

Besides doing all of this, I do not see the same output as I get locally, which is showing the results of the tests. Instead, all I can see is:

> running: node jasmine --config=content/examples/comparing-observables/jasmine.json
> completed: node jasmine --config=content/examples/comparing-observables/jasmine.json

Any help is appreciated.

mary-poppins commented:
You can preview 98e9ff4 at https://pr34537-98e9ff4.ngbuilds.io/.

mary-poppins commented:
You can preview a12fd79 at https://pr34537-a12fd79.ngbuilds.io/.

mary-poppins commented:
You can preview 05306e8 at https://pr34537-05306e8.ngbuilds.io/.

mary-poppins commented:
You can preview e167220 at https://pr34537-e167220.ngbuilds.io/.

mary-poppins commented:
You can preview d9cf686 at https://pr34537-d9cf686.ngbuilds.io/.

mary-poppins commented:
You can preview e956b8c at https://pr34537-e956b8c.ngbuilds.io/.

mary-poppins commented:
You can preview f4491cf at https://pr34537-f4491cf.ngbuilds.io/.

mary-poppins commented:
You can preview f040332 at https://pr34537-f040332.ngbuilds.io/.

mary-poppins commented:
You can preview 9a96eb2 at https://pr34537-9a96eb2.ngbuilds.io/.

mary-poppins commented:
You can preview 462e8b8 at https://pr34537-462e8b8.ngbuilds.io/.

mary-poppins commented:
You can preview 8ab4fe6 at https://pr34537-8ab4fe6.ngbuilds.io/.

sonukapoor commented:
@gkalpak I was able to run the unit tests in the `CI` environment. I believe this completes everything. Can you please review this one more time? 

![image](https://user-images.githubusercontent.com/59734/74697187-e32de900-51c7-11ea-9c31-711dac2435d0.png)

sonukapoor commented:
Blocked by https://github.com/angular/angular/pull/38143

mary-poppins commented:
You can preview 490eec2 at https://pr34537-490eec2.ngbuilds.io/.

mary-poppins commented:
You can preview 9f71b16 at https://pr34537-9f71b16.ngbuilds.io/.

mary-poppins commented:
You can preview 418e36f at https://pr34537-418e36f.ngbuilds.io/.

mary-poppins commented:
You can preview 23959b7 at https://pr34537-23959b7.ngbuilds.io/.

gkalpak reviewed (changes requested):
The tests look fine (with some minor comments). Thx for sorting this out.
I am not a big fan of the current setup. I'll look into whether it is possible to simplify it a bit.

sonukapoor committed:
test(docs-infra): add missing tests for obserables and promises

This commit adds missing tests for obserables and promises which
are both stand-alone mini-apps.

sonukapoor committed:
test(docs-infra): add commands to run the unit tests

This commit adds the necessary custom commands to run the tests in a
node environment.

gkalpak commented on aio/content/examples/comparing-observables/src/promises.ts:
```diff
@@ -1,25 +1,43 @@
+export function docRegionPromise(console: Console) {
+  // #docregion promise
+  // initiate execution
+  const promise = new Promise<number>((resolve, reject) => {
+    // Executer fn...
+    // #enddocregion promise
+    // The below is used in the unit tests.
+    resolve(2);
+    // #docregion promise
+  });
+
+  promise.then(value => {
+    // handle result here
+    // #enddocregion promise
+    // The below is used in the unit tests.
+    console.log(value);
+    // #docregion promise
+  });
+
+  // #enddocregion promise
+
+  // #docregion chain
+
+  promise.then(v => 2 * v);
+
+  // #enddocregion chain
```
I just realized that in order for this to be testable, we need to refactor the code a bit (and also use the "trick" from `docRegionError()` below).

```suggestion
  let promise = new Promise<number>((resolve, reject) => {
    // Executer fn...
    // #enddocregion promise
    // The below is used in the unit tests.
    resolve(2);
    // #docregion promise
  });

  // #enddocregion promise
  promise =
  // #docregion chain

  promise.then(v => 2 * v);

  // #enddocregion chain
  promise =
  // docregion promise
  promise.then(value => {
    // handle result here
    // #enddocregion promise
    // The below is used in the unit tests.
    console.log(value);
    // #docregion promise
  });

  // #enddocregion promise
  return promise;
```

[1]: https://github.com/angular/angular/blob/82d200a26f363f564a1f743bac92aa132638c7b3/aio/content/examples/comparing-observables/src/promises.ts#L33-L39

sonukapoor replied:
I guess at line 28, we need to return `value` as well, otherwise, the test will not pass, right?

gkalpak replied:
I have reordered the `then()` calls, so you can now assert that `console.log()` is called with `4` (i.e. the transformed value). You can also update the `docRegion...` function to accept the value to resolve the initial promise with.

mary-poppins commented:
You can preview d84dcc1 at https://pr34537-d84dcc1.ngbuilds.io/.

mary-poppins commented:
You can preview d4da011 at https://pr34537-d4da011.ngbuilds.io/.

gkalpak commented on aio/content/examples/comparing-observables/src/promises.ts:
```diff
@@ -1,25 +1,43 @@
+    // #enddocregion promise
+    // The below is used in the unit tests.
+    console.log(value);
+    return value;
```
Since you are returning the value here, there is little point in moving the `chain docregion above the `console.log()` block. Let's keep it below the `console.log()` block to avoid interleaving docregions.

mary-poppins commented:
You can preview 157a0f8 at https://pr34537-157a0f8.ngbuilds.io/.

gkalpak reviewed (changes requested):
The changes lgtm. Thx for the patience while we were figuring things out, @sonukapoor :grin: :100:

One last thing:
The `docs(docs-infra)` commit message looks wrong. It should either be squashed into the first commit or it should be a `test(docs-infra)` commit.

sonukapoor commented:
> The changes lgtm. Thx for the patience while we were figuring things out, @sonukapoor 😁 💯

No worries @gkalpak 

> One last thing:
> The `docs(docs-infra)` commit message looks wrong. It should either be squashed into the first commit or it should be a `test(docs-infra)` commit.

I have changed this squashed them. Please review one last time. Thanks

mary-poppins commented:
You can preview f3bd82e at https://pr34537-f3bd82e.ngbuilds.io/.

mary-poppins commented:
You can preview c265a4f at https://pr34537-c265a4f.ngbuilds.io/.

gkalpak commented on aio/content/examples/comparing-observables/src/promises.spec.ts:
```diff
@@ -0,0 +1,23 @@
+    pr.then((value) => {
+      expect(consoleLogSpy).toHaveBeenCalledTimes(1);
+      expect(consoleLogSpy).toHaveBeenCalledWith(2);
+      expect(value).toHaveBeenCalledWith(4);
```
You seem to have accidentally changed this while shifting commits around:

```suggestion
      expect(value).toBe(4);
```

sonukapoor replied:
Sorry about that. I think it should be fixed now.

mary-poppins commented:
You can preview e25f686 at https://pr34537-e25f686.ngbuilds.io/.

sonukapoor committed:
fixup! test(docs-infra): add missing tests for obserables and promises

mary-poppins commented:
You can preview ec9406f at https://pr34537-ec9406f.ngbuilds.io/.

gkalpak reviewed (approved):
Nice job, @sonukapoor! Thx :+1:

Reviewed-for: global-docs-approvers
